src/firejail/template.c

```diff
@@ -0,0 +1,429 @@
+								tmpl->value);
+
+	free(tmpl->key);
+	free(tmpl->value);
+	free(tmpl);
+}
+
+static Template* template_get(const char *key, int *err)
+{
+	Template *iter;
+	int key_len;
+
+	if (!tmpl_list) {
+		*err = -ENOENT;
+		return NULL;
+	}
+
+	if (!key || !*key) {
+		*err = -EINVAL;
+		return NULL;
+	}
+
+	key_len = strlen(key);
+	iter = tmpl_list;
+	while (iter) {
+		if (!strncmp(key, iter->key, key_len))
+			return iter;
+
+		iter = iter->next;
+	}
+
+	*err = -ENOKEY;
+	return NULL;
+}
+
+static const char* template_get_value(const char *key, int *err)
+{
+	Template *tmpl = template_get(key, err);
+
+	if (!tmpl)
+		return NULL;
+
+	return tmpl->value;
+}
+
+/* Prepend template to the list */
+static int template_add(Template *tmpl)
+{
+	int err;
+
+	if (!tmpl)
+		return -EINVAL;
+
+	if (!tmpl_list) {
+		tmpl_list = tmpl;
+		return 0;
+	}
+
+	if (template_get(tmpl->key, &err))
+		return -EALREADY;
+
+	tmpl->next = tmpl_list;
+	tmpl_list = tmpl;
+
+	return 0;
+}
+
+void template_cleanup()
+{
+	Template *iter;
+	Template *curr;
+
+	if (!tmpl_list)
+		return;
+
+	iter = tmpl_list;
+	while (iter) {
+		curr = iter;
+		iter = iter->next;
+		template_free(curr);
+	}
+
+	tmpl_list = NULL;
+}
+
+void template_print_all()
+{
+	Template *iter;
+
+	if (!tmpl_list)
+		return;
+
+	iter = tmpl_list;
+	while (iter) {
+		fprintf(stdout, "template key \"%s\" value \"%s\"\n",
+						iter->key, iter->value);
+		iter = iter->next;
+	}
+}
+
+static int is_valid_str(const char *str, StrCheckType type)
+{
+	int i;
+
+	if (!str || !*str)
+		return -EINVAL;
+
+	/* Keys or the values must start with an alphabetic char and must not
+	 * exceed D-Bus limit - 3 for name. As the template cannot be as first
+	 * minimum of 3 is reserved for countrycode/org/com/net/etc.
+	 */
+	if (!isalpha(*str) ||
+			(type == STR_CHECK_DBUS && strlen(str) > 0x000000FC))
+		return -EINVAL;
+
+	for (i = 1; str[i] != '\0'; i++)
+	{
+		if (iscntrl(str[i]))
+			return -EINVAL;
+
+		switch (type) {
+		case STR_CHECK_ALNUM:
+			if (!isalnum(str[i]))
+				return -EINVAL;
+
+			break;
+		case STR_CHECK_DBUS:
+			/* Allow '_./' in str */
+			if (!isalnum(str[i]) && str[i] != '_' &&
+					str[i] != '.' && str[i] != '/')
+				return -EINVAL;
+
+			/* Do not allow '..', prevents also '../'.*/
+			if (str[i] == '.' && str[i+1] == '.')
+				return -EINVAL;
+
+			break;
+		}
+	}
+
+	return 0;
+}
+
+void check_template(const char *arg) {
+	Template *tmpl;
+	const char *key;
+	const char *value;
+	char *delim = TEMPLATE_KEY_VALUE_DELIM;
+	char *macro = NULL;
+	char *macro_value = NULL;
+	int err = 0;
+
+	fprintf(stdout, "Check %s list %p\n", arg, tmpl_list);
+
+	/* Only alphanumeric chars in template key. */
+	key = strtok((char*)arg, delim);
+	if (is_valid_str(key, STR_CHECK_ALNUM) == -EINVAL) {
+		fprintf(stderr, "Error invalid template key \"%s\"", key);
+		exit(1);
+	}
+
+	/* Only a-zA-Z0-9_ */
+	value = strtok(NULL, delim);
+	if (is_valid_str(value, STR_CHECK_DBUS) == -EINVAL) {
```

Comment:
Because the values can be used in the D-Bus names in profile/permissions as well, so I think they should follow the syntax there. But apparently that `dbus_check_name()` could be used instead.

Reply:
I'm more wondering about: if this is general mechanism that is not necessarily related only to dbus names and such, would requirement of dbus purity already at this level prohibit some otherwise useful ways to utilize this functionality? Would it be better / are there already checks in places where things are used in dbus related activity?